### PR TITLE
Convert LDAP settings pages to React

### LINKED ISSUE #1621: Convert LDAP settings page to react

### PR DESCRIPTION
**Pull request by edmundoa**
This PR contains all necessary changes to make the LDAP settings and LDAP group mappings pages work again.

Sorry about the long list of commits, somehow got out of hand :confused: 

Fixes #1621.

### CODE REVIEW

File: graylog2-web-interface/public/stylesheets/graylog2.less [hunk 1]
```diff
@@ -1312,6 +1312,21 @@ td.limited {
   padding-top: 1px;
 }
 
+.ldap-test-login-result {
```
Comment (dennisoelkers) [inline graylog2-web-interface/public/stylesheets/graylog2.less]: Maybe it makes sense to defines these classes in a separate style and load/bundle it using the style-loader? This way we would start to bundle styles together with the components that use them and we could decouple/decompose the big global less file.

File: graylog2-web-interface/src/components/users/LdapComponent.jsx [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+      ldapUri: undefined,
+      roles: undefined,
+    };
+  },
+
+  componentDidMount() {
+    RolesStore.loadRoles().then(roles => {
+      this.setState({
+        roles: roles
+          .sort((r1, r2) => r1.name.localeCompare(r2.name))
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/users/LdapComponent.jsx]: Could you please extract the logic processing the roles into a separate function? This way we would have to change less when we change the RolesStore to reflux.

File: graylog2-web-interface/src/pages/LdapPage.jsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import DocsHelper from 'util/DocsHelper';
+
+import CurrentUserStore from 'stores/users/CurrentUserStore';
+
+import { PageHeader } from 'components/common';
+import DocumentationLink from 'components/support/DocumentationLink';
+import LdapComponent from 'components/users/LdapComponent';
+
+const LdapPage = React.createClass({
+  mixins: [Reflux.connect(CurrentUserStore)],
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/pages/LdapPage.jsx]: Why do we need this store in the page?

File: graylog2-web-interface/src/stores/users/LdapGroupsStore.ts [hunk 4]
```diff
@@ -1,49 +1,47 @@
...
+import jsRoutes = require('routing/jsRoutes');
+const fetch = require('logic/rest/FetchProvider').default;
 const UserNotification = require('util/UserNotification');
 const URLUtils = require('util/URLUtils');
 
+const LdapGroupsStore = {
+  loadGroups(): Promise<string[]> {
+    const promise = fetch('GET', URLUtils.qualifyUrl(jsRoutes.controllers.LdapController.apiGroups().url));
+    promise.catch(error => {
+      if (error.additional.status !== 404) {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/stores/users/LdapGroupsStore.ts]: What happens when the result is 404?

File: graylog2-web-interface/src/stores/users/LdapGroupsStore.ts [hunk 5]
```diff
@@ -1,49 +1,47 @@
...
+          'Could not load LDAP group list');
+      }
+    });
+
+    return promise;
+  },
+  loadMapping(): Promise<Object> {
+    const promise = fetch('GET', URLUtils.qualifyUrl(jsRoutes.controllers.LdapController.apiGroupsMapping().url));
+    promise.catch(error => {
+      if (error.additional.status !== 404) {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/stores/users/LdapGroupsStore.ts]: Same here, what happens when the result code is 404?

File: graylog2-web-interface/src/components/users/LdapComponent.jsx [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+  _setAdditionalDefaultGroups(rolesString) {
+    // only keep non-empty entries
+    const roles = rolesString.split(',').filter((v) => v !== '');
+    this._setSetting('additional_default_groups', roles);
+  },
+
+  _saveSettings(event) {
+    event.preventDefault();
+    LdapStore.update(this.state.ldapSettings);
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/users/LdapComponent.jsx]: Why is the store called directly and not decoupled via an action?

File: graylog2-web-interface/src/components/users/TestLdapConnection.jsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    // Reset connection status if ldapSettings changed
+    if (JSON.stringify(this.props.ldapSettings) !== JSON.stringify(nextProps.ldapSettings)) {
+      this.setState({serverConnectionStatus: {}});
+    }
+  },
+
+  _testServerConnection() {
+    LdapStore.testServerConnection(this.props.ldapSettings)
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/users/TestLdapConnection.jsx]: Why is the store called directly and not decoupled via an action?

**Comment by dennisoelkers** [comment]
I know that this was a huge change and very tedious to implement and I am very glad you did that, but wouldn't it make sense to rewrite the Stores that are touched in this PR to Reflux stores? Or is it too much work for this already very big changeset?

File: graylog2-web-interface/src/components/users/LdapComponent.jsx (hunk 2 above)
Comment (edmundoa) [inline graylog2-web-interface/src/components/users/LdapComponent.jsx]: Makes sense :ballot_box_with_check:

File: graylog2-web-interface/src/pages/LdapPage.jsx (hunk 3 above)
Comment (edmundoa) [inline graylog2-web-interface/src/pages/LdapPage.jsx]: As far as I know, we don't. Cleaning it up!

File: graylog2-web-interface/src/stores/users/LdapGroupsStore.ts (hunk 4 above)
Comment (edmundoa) [inline graylog2-web-interface/src/stores/users/LdapGroupsStore.ts]: Good question, the check was already in the store and I simply adapted it. I don't see anything special in the backend code, so I think we can remove it.

File: graylog2-web-interface/src/stores/users/LdapGroupsStore.ts (hunk 5 above)
Comment (edmundoa) [inline graylog2-web-interface/src/stores/users/LdapGroupsStore.ts]: Same as the other one.

File: graylog2-web-interface/src/components/users/LdapComponent.jsx (hunk 6 above)
Comment (edmundoa) [inline graylog2-web-interface/src/components/users/LdapComponent.jsx]: No good reason, fixed it :+1:

File: graylog2-web-interface/public/stylesheets/graylog2.less (hunk 1 above)
Comment (edmundoa) [inline graylog2-web-interface/public/stylesheets/graylog2.less]: Done! For now I put it in the same level as the component, as I think it's easier to see when files move or disappear, but I'm open for feedback on that.

**Comment by edmundoa** [comment]
As I mentioned in Slack, the reason for not migrating the stores is that they didn't need much work to fix them (basically changing the promises and routes), and the list of changes was indeed large. I will migrate the `LdapGroupsStore`, as it's only used in the files already touched in this PR, but will leave the `RolesStore` as is for now.

**Pull request merged**